ecosystem/purescript-0.14-migration-guide.md

```diff
@@ -0,0 +1,622 @@
+
+```txt
+Data.Generic.Rep.Bounded        -> Data.Bounded.Generic
+Data.Generic.Rep.Enum           -> Data.Enum.Generic (in the `enums` package)
+Data.Generic.Rep.Eq             -> Data.Eq.Generic
+Data.Generic.Rep.HeytingAlgebra -> Data.HeytingAlgebra.Generic
+Data.Generic.Rep.Monoid         -> Data.Monoid.Generic
+Data.Generic.Rep.Ord            -> Data.Ord.Generic
+Data.Generic.Rep.Ring           -> Data.Ring.Generic
+Data.Generic.Rep.Semigroup      -> Data.Semigroup.Generic
+Data.Generic.Rep.Semiring       -> Data.Semiring.Generic
+Data.Generic.Rep.Show           -> Data.Show.Generic
+```
+
+The `argonaut-generic` library has also renamed its modules to match this naming convention:
+
+```txt
+Data.Argonaut.Decode.Generic.Rep -> Data.Argonaut.Decode.Generic
+Data.Argonaut.Encode.Generic.Rep -> Data.Argonaut.Encode.Generic Data.Argonaut.Types.Generic.Rep  -> Data.Argonaut.Types.Generic
+```
+
+#### The `globals` library has been deprecated
+
+The `globals` library provided various functions and values that are in the global scope in JavaScript. The library has been deprecated in particular because PureScript supports multiple backends besides JavaScript, and the relevant functions have been moved or reimplemented in other libraries.
+
+Specifically:
+
+* Number values and operations like `isFinite`, parsing, and formatting have been moved to the `numbers` package, which itself has moved into the core organization.
+* Integer parsing has been removed, as it is already included in the `integers` package.
+* Encoding and decoding URIs and URI components using JavaScript's builtin `encodeURIComponent` and `decodeURIComponent` have been moved to the new `js-uri` package.
+* Unsafe global functions have been removed altogether.
+
+**To fix:**
+
+You should remove the `globals` dependency from your library or application.
+
+If you used any `Number`-related code from the `globals` package, then you should add a new dependency on `numbers` and rename your imports:
+
+```diff
+- import Global (isNan, nan, isFinite, infinity)
++ import Data.Number (isNan, nan, isFinite, infinity)
+ 
+- import Global (toFixed, toPrecision, toExponential)
++ import Data.Number.Format (toStringWith, fixed, precision, exponential)
+
+  -- toFixed, toPrecision, and toExponential can be replaced as follows:
+- x = toFixed someInt 4.0 
++ x = toStringWith (fixed someInt) 4.0
+ 
+- y = toPrecision someInt 4.0
++ y = toStringWith (precision someInt) 4.0
+ 
+- z = toExponential intArg 4.0
++ z = toStringWith (exponential someInt) 4.0
+```
+
+If you used `readInt`, then you should add a dependency on `integers` and use the `fromStringAs` function given the correct radix, or if you are parsing a base-10 integer then you can use `Data.Number.fromString`.
+
+```diff
+- import Global (readInt)
++ import Data.Int (radix, fromStringAs)
+
+  x :: String -> Maybe Int
+- x = readInt 2
++ x = fromStringAs (unsafeRadix 2)
++   where
++   unsafeRadix n = unsafePartial fromJust $ radix n
```

Comment:
Sorry to go off topic, but I think we should provide a Radix constructor which always succeeds by clamping the input to the valid range, since this is kind of gross. The `numbers` library already clamps in the `exponential`, `fixed` etc formatters.

Reply:
Perfectly happy with that myself! This is just needed in the meantime.

Reply:
I made https://github.com/purescript/purescript-integers/issues/46

Reply:
I just realised that Data.Int provides `binary :: Radix`, so this could be more nicely written as `fromStringAs binary`?

Reply:
Yes, that's better. Just saw that myself, good catch!